cabal-install/Distribution/Client/Dependency/TopDown.hs

```diff
@@ -669,8 +671,9 @@ finaliseSelectedPackages pref selected constraints =
         -- version constraints. TODO: distinguish hacks from prefs
         bounded = boundedAbove versionRange
         isPreferred p
-          | bounded   = True -- any constant will do
-          | otherwise = packageVersion p `withinRange` preferredVersions
+          | bounded   = 999 -- any constant will do
```

Comment:
I'm not familiar with the Topdown solver, but it looks like this function ignores preferred ranges when the dependency is bounded above.  Since `bounded` doesn't depend on the package instances being compared, the value here will only be compared with itself and should have no effect on behavior (which is consistent with the comment).  I would use something like `0` or `1`, so that it doesn't look like the constant has special meaning.  I guess moving the guard into `bestByPref` would remove the need for an arbitrary value, but I don't recommend refactoring it now that the solver is deprecated.